### DISCUSSION TITLE
👋 - newbie here. Trying to integrate static library; works but has some odd results.

### DISCUSSION THREAD

**Discussion by sortofsleepy** [c3lang/c3c#1476] [Q&A]
Hello!

New to c3 here, looks pretty interesting so far. 

I thought I'd try to wrap my head around things by building a small toy renderer with [wgpu-native](https://github.com/gfx-rs/wgpu-native/). 

Apologies if this has been discussed but I wasn't sure what to search for - I'm running into an odd issue where for whatever reason, when I call functions in the library, it seemingly breaks something as no print output aside from the launch message(after running clean-run) is produced and no exit code is mentioned. 

If I leave out the native function calls, the print function works and the console reports the exit code. 

Just curious as to what might be going on, being able to print would be nice haha. 

I can upload a demo repo if it's helpful but given that the library size is a bit big, figured it'd be better to start without in case it's a simple thing. 

Lastly
 
* I am on Windows and the compiler version is 0.6.3.
* In case it's not clear, I'm running off of the default project template and calling`c3c clean-run` 

EDIT 
===
I'm starting to suspect this is library related; tried a barebones static c library that just prints hello world and that seems to work fine so I believe my syntax is correct. Feel free to close this topic but just in case there's something else that could be the issue, it'd be great to leave this up for at least a few days. 

Thank you for any ideas. 

Here is the current code

```c

/// app 

module myapp;
import std::io;
import wgpu;


fn int main(String[] args)
{
    wgpu::InstanceDescriptor desc = { null };
	wgpu::Instance ctx = wgpu::createInstance(&desc);

    wgpu::release(ctx);

    std::io::printn("Hello, World!");

	return 0;
}


```

Here's the current interface 
```

// wrapper into wgpu-native

module wgpu;
def Instance = void*;

struct InstanceDescriptor {
    void* nextInChain;
}

fn Instance createInstance(InstanceDescriptor* descriptor = null) @extern("wgpuCreateInstance");
fn void release(Instance instance) @extern("wgpuInstanceRelease");

```

Lastly, here's the current project json 

```json 
{
  // Language version of C3.
  "langrev": "1",
  // Warnings used for all targets.
  "warnings": [
    "no-unused"
  ],
  "linked-libraries": [
    "lib/wgpu/wgpu_native.dll",
    "lib/wgpu/wgpu_native",
  ],
  // Directories where C3 library files may be found.
  "dependency-search-paths": [
    "lib"
  ],
  // Libraries to use for all targets.
  "dependencies": [],
  // Authors, optionally with email.
  "authors": [
    "John Doe <john.doe@example.com>"
  ],
  // Version using semantic versioning.
  "version": "0.1.0",
  // Sources compiled for all targets.
  "sources": [
    "src/**",
    "src/wgpu/**"
  ],
  // C sources if the project also compiles C sources
  // relative to the project file.
  // "c-sources": [ "csource/**" ],
  // Include directories for C sources relative to the project file.
  // "c-include-dirs": [ "csource/include" ],
  // Output location, relative to project file.
  "output": "build",
  // Architecture and OS target.
  // You can use 'c3c --list-targets' to list all valid targets.
  // "target": "windows-x64",
  // Targets.
  "targets": {
    "myapp": {
      // Executable or library.
      "type": "executable"
      // Additional libraries, sources
      // and overrides of global settings here.
    }
  },
  // Global settings.
  // CPU name, used for optimizations in the LLVM backend.
  "cpu": "generic",
  // Optimization: "O0", "O1", "O2", "O3", "O4", "O5", "Os", "Oz".
  "opt": "O0"
  // See resources/examples/project_all_settings.json and 'c3c --list-project-properties' to see more properties.
}

```

**Comment by sortofsleepy** [comment]
Ok, I'm an idiot. 

Forgot to add dll to executable directory haha :(

**Comment by sortofsleepy** [comment]
Sorry forgot to add; I was only reminded of needing to add .dlls to the same directory as the executable after setting up a simple pure C example; Windows brought up an alert saying a dll was missing. 

I don't know how easy it would be to add but I think it would be helpful to have a similar message{with or without a GUI window) when something fails to run as the current system of outputting nothing is a bit difficult to troubleshoot.

**Reply by lerno** [reply]
On Windows, were you doing a command line app or a windows app? I was thinking that maybe it's something different depending on whether one is using a windows based app or not.

Btw, you can mark your app `@winmain` to make the app run without a console to test if that would work give a popup in the missing DLL case.

**Reply by sortofsleepy** [reply]
> On Windows, were you doing a command line app or a windows app? I was thinking that maybe it's something different depending on whether one is using a windows based app or not.
> 
> Btw, you can mark your app `@winmain` to make the app run without a console to test if that would work give a popup in the missing DLL case.


I'm just making a command line app. 

The test app(that mentioned dlls were missing) was also command line based; I am using CLion and used the default project template for a C executable application. It does use CMake to build; perhaps something about how the project is setup with CMake is causing something to load behind the scenes in order to surface any error messages?

As for adding `@winmain`, it doesn't bring up a popup window like the test app.

**Reply by lerno** [reply]
So what gave you the popup? Windows or CMake?

**Reply by sortofsleepy** [reply]
That I don't know unfortunately. The pop up window definitely was a GUI window so presumably it was coming from Windows, but I don't know what specifically could have triggered it. Like I said, I just used the default project template and didn't make any changes aside from adding the wgpu libraries/include folder. 

If it helps, this is the CMakeLists.txt. If there's something going on behind the scenes that CLion is doing I don't know what that is. 

```
cmake_minimum_required(VERSION 3.28)
project(testlib)

set(CMAKE_CXX_STANDARD 17)

add_executable(Test main.cpp)

target_link_libraries(Test "${PROJECT_SOURCE_DIR}/wgpu/wgpu_native.dll.lib")
target_include_directories(Test INTERFACE "${PROJECT_SOURCE_DIR}/wgpu/include")
```

**Reply by lerno** [reply]
It might be that there is no window yet popped: https://github.com/microsoft/terminal/issues/9788

**Reply by sortofsleepy** [reply]
Just did a quick test after thinking about this a little more. 

So in CLion, like other IDEs, there's a way to build/launch your application within it's gui which is what I normally would do vs running things from a terminal. Running it through the CLion gui seems to allow the missing dll popup to appear. 

If I run the built exe of the test project from the command line I get no alert like with the c3 version. 

I don't know if it has to do with windowing or not like in the issue you mentioned; the project "should" be a command line one, but seems like CLion might have had something to do with why I got a warning. 

Anyways seems like there probably isn't any issue so forget what I said haha. 

Thanks for responding and a big thank you to you and the other contributors for building this.

**Reply by lerno** [reply]
I think you can still file an issue on this. Maybe someone can pick it up and see if there is a way to improve it.

**Reply by sortofsleepy** [reply]
Sounds good, I'll do that. It would be nice if someone could figure it out.
